common/persistence/sql/sqlHistoryV2Manager.go

```diff
@@ -0,0 +1,502 @@
+	treeID := *branch.TreeID
+	brsToDelete := branch.Ancestors
+	beginNodeID := p.GetBeginNodeID(branch)
+	brsToDelete = append(brsToDelete, &shared.HistoryBranchRange{
+		BranchID:    branch.BranchID,
+		BeginNodeID: common.Int64Ptr(beginNodeID),
+	})
+
+	rsp, err := m.GetHistoryTree(&p.GetHistoryTreeRequest{
+		TreeID: treeID,
+	})
+	if err != nil {
+		return err
+	}
+	// We won't delete the branch if there is any branch forking in progress. We will return error.
+	if len(rsp.ForkingInProgressBranches) > 0 {
+		return &p.ConditionFailedError{
+			Msg: fmt.Sprintf("There are branches in progress of forking"),
+		}
+	}
+
+	// If there is no branch forking in progress we see here, it means that we are safe to calculate the deleting ranges based on the current result,
+	// Because before getting here, we've already deleted mutableState record, so all the forking branches in the future should fail.
+
+	// validBRsMaxEndNode is to for each branch range that is being used, we want to know what is the max nodeID referred by other valid branch
+	validBRsMaxEndNode := map[string]int64{}
+	for _, b := range rsp.Branches {
+		for _, br := range b.Ancestors {
+			curr, ok := validBRsMaxEndNode[*br.BranchID]
+			if !ok || curr < *br.EndNodeID {
+				validBRsMaxEndNode[*br.BranchID] = *br.EndNodeID
+			}
+		}
+	}
+
+	return m.txExecute("DeleteHistoryBranch", func(tx sqldb.Tx) error {
+		branchID := sqldb.MustParseUUID(*branch.BranchID)
+		treeFilter := &sqldb.HistoryTreeFilter{
+			TreeID:   sqldb.MustParseUUID(treeID),
+			BranchID: &branchID,
+		}
+		_, err := tx.DeleteFromHistoryTree(treeFilter)
+		if err != nil {
+			return err
+		}
+
+		done := false
+		// for each branch range to delete, we iterate from bottom to up, and delete up to the point according to validBRsEndNode
+		for i := len(brsToDelete) - 1; i >= 0; i-- {
+			br := brsToDelete[i]
+			maxReferredEndNodeID, ok := validBRsMaxEndNode[*br.BranchID]
+			nodeFilter := &sqldb.HistoryNodeFilter{
+				TreeID:   sqldb.MustParseUUID(treeID),
+				BranchID: sqldb.MustParseUUID(*br.BranchID),
+			}
+
+			if ok {
+				// we can only delete from the maxEndNode and stop here
+				nodeFilter.MinNodeID = &maxReferredEndNodeID
+				done = true
+			} else {
+				// No any branch is using this range, we can delete all of it
+				nodeFilter.MinNodeID = br.BeginNodeID
+			}
+			_, err := tx.DeleteFromHistoryNode(nodeFilter)
+			if err != nil {
+				return err
+			}
+			if done {
+				break
+			}
+		}
+		return nil
+	})
+}
+
+// UpdateHistoryBranch update a branch
+func (m *sqlHistoryV2Manager) CompleteForkBranch(request *p.InternalCompleteForkBranchRequest) error {
+	branch := request.BranchInfo
+	treeID := sqldb.MustParseUUID(*branch.TreeID)
+	branchID := sqldb.MustParseUUID(*branch.BranchID)
+	if request.Success {
+		row := &sqldb.HistoryTreeRow{
+			TreeID:     treeID,
+			BranchID:   branchID,
+			InProgress: false,
+		}
+		result, err := m.db.UpdateHistoryTree(row)
+		if err != nil {
+			return err
```

Comment:
wrap this in InternalServiceError - if you want the caller to retry on this error